File: Sources/RPC/Envelope.swift
import Foundation

public enum RPCProtocol: String, Sendable, Hashable, CaseIterable, Codable {
  case daemon
  case worker
  case guest

  /// Wire version implemented by this module.
  public static let currentVersion = 1

  /// Maximum single frame size accepted on this channel.
  public var frameCap: Int {
    switch self {
    case .daemon: return 16 * 1024 * 1024
    case .worker, .guest: return 4 * 1024 * 1024
    }
  }
}

public enum EnvelopeKind: String, Sendable, Hashable, CaseIterable, Codable {
  case request
  case response
  case event
  case chunk
  case cancel
}

public struct Envelope: Sendable, Hashable {
  public var protocolName: RPCProtocol
  public var version: Int
  public var kind: EnvelopeKind
  public var requestId: String
  public var method: String?
  public var streamSeq: Int64?
  public var end: Bool?
  public var payload: JSONValue?
  public var error: RPCErrorPayload?

  public init(
    protocolName: RPCProtocol,
    version: Int = RPCProtocol.currentVersion,
    kind: EnvelopeKind,
    requestId: String,
    method: String? = nil,
    streamSeq: Int64? = nil,
    end: Bool? = nil,
    payload: JSONValue? = nil,
    error: RPCErrorPayload? = nil
  ) {
    self.protocolName = protocolName
    self.version = version
    self.kind = kind
    self.requestId = requestId
    self.method = method
    self.streamSeq = streamSeq
    self.end = end
    self.payload = payload
    self.error = error
  }
}

// MARK: - Factories

extension Envelope {
  public static func request(
    _ proto: RPCProtocol, requestId: String, method: String, payload: JSONValue? = nil
  ) -> Envelope {
    Envelope(protocolName: proto, kind: .request, requestId: requestId, method: method, payload: payload)
  }

  public static func event(
    _ proto: RPCProtocol, requestId: String, method: String, payload: JSONValue? = nil
  ) -> Envelope {
    Envelope(protocolName: proto, kind: .event, requestId: requestId, method: method, payload: payload)
  }

  public static func response(
    _ proto: RPCProtocol, requestId: String, method: String? = nil, payload: JSONValue
  ) -> Envelope {
    Envelope(protocolName: proto, kind: .response, requestId: requestId, method: method, payload: payload)
  }

  public static func failure(
    _ proto: RPCProtocol, requestId: String, method: String? = nil, error: RPCErrorPayload
  ) -> Envelope {
    Envelope(protocolName: proto, kind: .response, requestId: requestId, method: method, error: error)
  }

  public static func chunk(
    _ proto: RPCProtocol, requestId: String, method: String? = nil, streamSeq: Int64, end: Bool,
    payload: JSONValue? = nil, error: RPCErrorPayload? = nil
  ) -> Envelope {
    Envelope(
      protocolName: proto, kind: .chunk, requestId: requestId, method: method,
      streamSeq: streamSeq, end: end, payload: payload, error: error)
  }

  public static func cancel(_ proto: RPCProtocol, requestId: String) -> Envelope {
    Envelope(protocolName: proto, kind: .cancel, requestId: requestId)
  }
}

// MARK: - Encoding

extension Envelope {
  static let allowedKeys: Set<String> = [
    "protocol", "version", "kind", "requestId", "method", "streamSeq", "end", "payload", "error",
  ]

  /// Deterministic: keys sorted, absent members omitted.
  public func encode() -> [UInt8] {
    var members: [String: JSONValue] = [
      "protocol": .string(protocolName.rawValue),
      "version": .int(Int64(version)),
      "kind": .string(kind.rawValue),
      "requestId": .string(requestId),
    ]
    if let method { members["method"] = .string(method) }
    if let streamSeq { members["streamSeq"] = .int(streamSeq) }
    if let end { members["end"] = .bool(end) }
    if let payload { members["payload"] = payload }
    if let error {
      members["error"] = .object([
        "code": .string(error.code),
        "message": .string(error.message),
        "retryable": .bool(error.retryable),
      ])
    }
    return JSONValue.object(members).encoded()
  }
}

// MARK: - Decoding

extension Envelope {
  public static func decode(
    from bytes: [UInt8], expecting proto: RPCProtocol
  ) throws(EnvelopeError) -> Envelope {
    var outcome: Result<Envelope, EnvelopeError>?
    bytes.withUnsafeBytes { buffer in
      do {
        outcome = .success(try Envelope.decode(from: buffer, expecting: proto))
      } catch let error as EnvelopeError {
        outcome = .failure(error)
      } catch {
        outcome = .failure(EnvelopeError(.malformed, "\(error)"))
      }
    }
    switch outcome! {
    case .success(let envelope): return envelope
    case .failure(let error): throw error
    }
  }

  public static func decode(
    from bytes: UnsafeRawBufferPointer, expecting proto: RPCProtocol
  ) throws(EnvelopeError) -> Envelope {
    let root: JSONValue
    do {
      root = try StrictJSON.parse(bytes)
    } catch {
      throw EnvelopeError(.malformed, "invalid JSON: \(error)")
    }
    guard case .object(let members) = root else {
      throw EnvelopeError(.malformed, "envelope must be a JSON object")
    }
    let salvagedId = members["requestId"]?.stringValue.flatMap { $0.isEmpty ? nil : $0 }
    return try validate(members, salvagedId: salvagedId, expecting: proto)
  }

  private static func validate(
    _ members: [String: JSONValue], salvagedId: String?, expecting proto: RPCProtocol
  ) throws(EnvelopeError) -> Envelope {
    for key in members.keys where !allowedKeys.contains(key) {
      throw EnvelopeError(.malformed, "unknown key '\(key)'", requestId: salvagedId)
    }
    guard let version = members["version"]?.intValue else {
      throw EnvelopeError(.malformed, "missing or non-integer 'version'", requestId: salvagedId)
    }
    guard version == Int64(RPCProtocol.currentVersion) else {
      throw EnvelopeError(.protocolVersion, "unsupported version \(version)", requestId: salvagedId)
    }
    guard let rawProtocol = members["protocol"]?.stringValue,
      let protocolName = RPCProtocol(rawValue: rawProtocol)
    else {
      throw EnvelopeError(.malformed, "missing or unknown 'protocol'", requestId: salvagedId)
    }
    guard protocolName == proto else {
      throw EnvelopeError(
        .protocolMismatch, "expected '\(proto.rawValue)', got '\(rawProtocol)'",
        requestId: salvagedId)
    }
    guard let rawKind = members["kind"]?.stringValue, let kind = EnvelopeKind(rawValue: rawKind) else {
      throw EnvelopeError(.malformed, "missing or unknown 'kind'", requestId: salvagedId)
    }
    guard let requestId = salvagedId else {
      throw EnvelopeError(.malformed, "missing or empty 'requestId'")
    }
    var envelope = Envelope(
      protocolName: protocolName, version: Int(version), kind: kind, requestId: requestId)
    try applyOptionalMembers(members, to: &envelope)
    try checkShape(envelope, members: members)
    return envelope
  }

  private static func applyOptionalMembers(
    _ members: [String: JSONValue], to envelope: inout Envelope
  ) throws(EnvelopeError) {
    let requestId = envelope.requestId
    if let raw = members["method"] {
      guard let method = raw.stringValue, !method.isEmpty else {
        throw EnvelopeError(.malformed, "'method' must be a non-empty string", requestId: requestId)
      }
      envelope.method = method
    }
    if let raw = members["streamSeq"] {
      guard let seq = raw.intValue, seq >= 0 else {
        throw EnvelopeError(.malformed, "'streamSeq' must be a non-negative integer", requestId: requestId)
      }
      envelope.streamSeq = seq
    }
    if let raw = members["end"] {
      guard let end = raw.boolValue else {
        throw EnvelopeError(.malformed, "'end' must be a boolean", requestId: requestId)
      }
      envelope.end = end
    }
    envelope.payload = members["payload"]
    if let raw = members["error"] {
      envelope.error = try decodeError(raw, requestId: requestId)
    }
  }

  private static func decodeError(
    _ raw: JSONValue, requestId: String
  ) throws(EnvelopeError) -> RPCErrorPayload {
    guard case .object(let fields) = raw else {
      throw EnvelopeError(.malformed, "'error' must be an object", requestId: requestId)
    }
    for key in fields.keys where !["code", "message", "retryable"].contains(key) {
      throw EnvelopeError(.malformed, "unknown 'error' key '\(key)'", requestId: requestId)
    }
    guard let code = fields["code"]?.stringValue, !code.isEmpty,
      let message = fields["message"]?.stringValue
    else {
      throw EnvelopeError(.malformed, "'error' needs 'code' and 'message'", requestId: requestId)
    }
    if let retryable = fields["retryable"], retryable.boolValue == nil {
      throw EnvelopeError(.malformed, "'error.retryable' must be a boolean", requestId: requestId)
    }
    return RPCErrorPayload(
      code: code, message: message, retryable: fields["retryable"]?.boolValue ?? false)
  }

  private static func checkShape(
    _ envelope: Envelope, members: [String: JSONValue]
  ) throws(EnvelopeError) {
    let requestId = envelope.requestId
    func reject(_ why: String) -> EnvelopeError {
      EnvelopeError(.malformed, why, requestId: requestId)
    }
    let hasStreamMembers = members["streamSeq"] != nil || members["end"] != nil
    switch envelope.kind {
    case .request, .event:
      guard envelope.method != nil else { throw reject("'\(envelope.kind.rawValue)' needs 'method'") }
      guard !hasStreamMembers else { throw reject("'\(envelope.kind.rawValue)' has stream members") }
      guard envelope.error == nil else { throw reject("'\(envelope.kind.rawValue)' has 'error'") }
    case .response:
      guard !hasStreamMembers else { throw reject("'response' has stream members") }
      guard (members["payload"] == nil) != (members["error"] == nil) else {
        throw reject("'response' needs exactly one of 'payload' or 'error'")
      }
    case .chunk:
      guard envelope.streamSeq != nil else { throw reject("'chunk' needs 'streamSeq'") }
      guard envelope.end != nil else { throw reject("'chunk' needs 'end'") }
    case .cancel:
      guard members["method"] == nil, !hasStreamMembers, members["payload"] == nil,
        members["error"] == nil
      else { throw reject("'cancel' carries only 'requestId'") }
    }
  }
}

File: GuestAgent/internal/rpc/envelope.go
// Package rpc implements the envelope framing protocol shared by the
// daemon, worker and guest channels, per ../../../Proto/envelope.md.
// Decoding is strict and defensive: the guest treats every inbound frame
// as untrusted input.
package rpc

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"io"
)

// Envelope kinds, per envelope.md.
const (
	KindRequest  = "request"
	KindResponse = "response"
	KindEvent    = "event"
	KindChunk    = "chunk"
	KindCancel   = "cancel"
)

// protocolVersion is the only version this package understands.
const protocolVersion = 1

// Envelope is the wire-level RPC message. JSON field names are exactly as
// specified in envelope.md; see Decode for the validation rules.
type Envelope struct {
	Protocol  string          `json:"protocol"`
	Version   int             `json:"version"`
	Kind      string          `json:"kind"`
	RequestID string          `json:"requestId"`
	Method    string          `json:"method,omitempty"`
	StreamSeq *int64          `json:"streamSeq,omitempty"`
	End       *bool           `json:"end,omitempty"`
	Payload   json.RawMessage `json:"payload,omitempty"`
	Error     *ErrorPayload   `json:"error,omitempty"`
}

// ErrorPayload is the wire shape of Envelope.Error.
type ErrorPayload struct {
	Code      string `json:"code"`
	Message   string `json:"message"`
	Retryable bool   `json:"retryable"`
}

// allowedEnvelopeFields is the complete set of top-level envelope keys.
// Anything else is rejected as MALFORMED (strict decoding).
var allowedEnvelopeFields = map[string]bool{
	"protocol":  true,
	"version":   true,
	"kind":      true,
	"requestId": true,
	"method":    true,
	"streamSeq": true,
	"end":       true,
	"payload":   true,
	"error":     true,
}

var validKinds = map[string]bool{
	KindRequest:  true,
	KindResponse: true,
	KindEvent:    true,
	KindChunk:    true,
	KindCancel:   true,
}

// Encode marshals an Envelope to its canonical JSON wire form. Encoding is
// deterministic: encoding/json always emits struct fields in declaration
// order, and json.RawMessage/json.Number payloads are copied byte-for-byte,
// so integers never round-trip through float64.
func Encode(e Envelope) ([]byte, error) {
	return json.Marshal(e)
}

// Decode parses and strictly validates a single envelope from b.
//
// Rejected as MALFORMED: empty input, non-object top level, unknown
// top-level keys, duplicate keys at any object nesting level, trailing
// bytes after the JSON value, and kind-specific shape violations (request/
// event without method, chunk without streamSeq+end, response without
// exactly one of payload/error, cancel with anything beyond requestId).
//
// version != 1 is rejected as PROTOCOL_VERSION. protocol != expectProtocol
// is rejected as PROTOCOL_MISMATCH. Both checks run only after the document
// is structurally well-formed.
func Decode(b []byte, expectProtocol string) (Envelope, error) {
	fields, err := decodeTopLevel(b)
	if err != nil {
		return Envelope{}, err
	}

	protocol, err := requireString(fields, "protocol")
	if err != nil {
		return Envelope{}, err
	}
	version, err := requireInt64(fields, "version")
	if err != nil {
		return Envelope{}, err
	}
	kind, err := requireString(fields, "kind")
	if err != nil {
		return Envelope{}, err
	}
	requestID, err := requireString(fields, "requestId")
	if err != nil {
		return Envelope{}, err
	}

	if version != protocolVersion {
		return Envelope{}, newRPCError(CodeProtocolVersion,
			fmt.Sprintf("unsupported protocol version %d", version))
	}
	if protocol != expectProtocol {
		return Envelope{}, newRPCError(CodeProtocolMismatch,
			fmt.Sprintf("expected protocol %q, got %q", expectProtocol, protocol))
	}
	if !validKinds[kind] {
		return Envelope{}, newMalformed(fmt.Sprintf("unknown kind %q", kind))
	}

	env := Envelope{Protocol: protocol, Version: int(version), Kind: kind, RequestID: requestID}

	if raw, ok := fields["method"]; ok {
		s, err := parseString(raw)
		if err != nil {
			return Envelope{}, newMalformed("field \"method\" must be a string")
		}
		env.Method = s
	}
	if raw, ok := fields["payload"]; ok {
		env.Payload = raw
	}
	if raw, ok := fields["error"]; ok {
		var ep ErrorPayload
		if err := strictUnmarshal(raw, &ep); err != nil {
			return Envelope{}, newMalformed("field \"error\" is invalid: " + err.Error())
		}
		env.Error = &ep
	}
	if raw, ok := fields["streamSeq"]; ok {
		seq, err := parseInt64(raw)
		if err != nil {
			return Envelope{}, newMalformed("field \"streamSeq\" must be an integer")
		}
		env.StreamSeq = &seq
	}
	if raw, ok := fields["end"]; ok {
		var end bool
		if err := json.Unmarshal(raw, &end); err != nil {
			return Envelope{}, newMalformed("field \"end\" must be a boolean")
		}
		env.End = &end
	}

	if err := validateKindShape(env); err != nil {
		return Envelope{}, err
	}
	return env, nil
}

// validateKindShape enforces the per-kind structural rules from
// envelope.md's "Rules" section.
func validateKindShape(env Envelope) error {
	switch env.Kind {
	case KindRequest, KindEvent:
		if env.Method == "" {
			return newMalformed(fmt.Sprintf("kind %q requires \"method\"", env.Kind))
		}
	case KindChunk:
		if env.StreamSeq == nil {
			return newMalformed("kind \"chunk\" requires \"streamSeq\"")
		}
		if env.End == nil {
			return newMalformed("kind \"chunk\" requires \"end\"")
		}
	case KindResponse:
		hasPayload := env.Payload != nil
		hasError := env.Error != nil
		if hasPayload == hasError {
			return newMalformed("kind \"response\" requires exactly one of payload or error")
		}
	case KindCancel:
		if env.Method != "" || env.Payload != nil || env.Error != nil || env.StreamSeq != nil || env.End != nil {
			return newMalformed("kind \"cancel\" must contain only requestId")
		}
	}
	return nil
}

// decodeTopLevel validates that data is exactly one well-formed JSON object
// with no duplicate keys at any nesting level, no unknown top-level keys,
// and no trailing bytes, returning the raw bytes of each top-level field.
func decodeTopLevel(data []byte) (map[string]json.RawMessage, error) {
	if len(bytes.TrimSpace(data)) == 0 {
		return nil, newMalformed("empty input")
	}

	dec := json.NewDecoder(bytes.NewReader(data))
	dec.UseNumber()

	tok, err := dec.Token()
	if err != nil {
		return nil, newMalformed("invalid JSON: " + err.Error())
	}
	delim, ok := tok.(json.Delim)
	if !ok || delim != '{' {
		return nil, newMalformed("top-level JSON value must be an object")
	}

	fields := make(map[string]json.RawMessage)
	for dec.More() {
		keyTok, err := dec.Token()
		if err != nil {
			return nil, newMalformed("invalid JSON: " + err.Error())
		}
		key, ok := keyTok.(string)
		if !ok {
			return nil, newMalformed("object key is not a string")
		}
		if !allowedEnvelopeFields[key] {
			return nil, newMalformed(fmt.Sprintf("unknown field %q", key))
		}
		if _, dup := fields[key]; dup {
			return nil, newMalformed(fmt.Sprintf("duplicate field %q", key))
		}

		// Token() and Decode() may be freely interleaved on the same
		// Decoder: this captures the raw bytes of the value that starts
		// at the decoder's current position without parsing numbers
		// into float64 (json.RawMessage copies bytes verbatim).
		var raw json.RawMessage
		if err := dec.Decode(&raw); err != nil {
			return nil, newMalformed("invalid JSON: " + err.Error())
		}
		if err := checkNoDuplicateKeys(raw); err != nil {
			return nil, err
		}
		fields[key] = raw
	}

	if _, err := dec.Token(); err != nil { // closing '}'
		return nil, newMalformed("invalid JSON: " + err.Error())
	}
	if _, err := dec.Token(); !errors.Is(err, io.EOF) {
		if err == nil {
			return nil, newMalformed("trailing data after envelope")
		}
		return nil, newMalformed("invalid JSON: " + err.Error())
	}

	return fields, nil
}

// checkNoDuplicateKeys recursively walks a raw JSON value and rejects any
// object — at any nesting level, including inside payload/error — that
// contains a repeated key.
func checkNoDuplicateKeys(raw json.RawMessage) error {
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	return walkNoDuplicates(dec)
}

func walkNoDuplicates(dec *json.Decoder) error {
	tok, err := dec.Token()
	if err != nil {
		return newMalformed("invalid JSON: " + err.Error())
	}
	delim, ok := tok.(json.Delim)
	if !ok {
		return nil // scalar: string, json.Number, bool, or null
	}

	switch delim {
	case '{':
		seen := make(map[string]struct{})
		for dec.More() {
			keyTok, err := dec.Token()
			if err != nil {
				return newMalformed("invalid JSON: " + err.Error())
			}
			key, ok := keyTok.(string)
			if !ok {
				return newMalformed("object key is not a string")
			}
			if _, dup := seen[key]; dup {
				return newMalformed(fmt.Sprintf("duplicate key %q", key))
			}
			seen[key] = struct{}{}
			if err := walkNoDuplicates(dec); err != nil {
				return err
			}
		}
		if _, err := dec.Token(); err != nil { // closing '}'
			return newMalformed("invalid JSON: " + err.Error())
		}
	case '[':
		for dec.More() {
			if err := walkNoDuplicates(dec); err != nil {
				return err
			}
		}
		if _, err := dec.Token(); err != nil { // closing ']'
			return newMalformed("invalid JSON: " + err.Error())
		}
	}
	return nil
}

// strictUnmarshal decodes raw into v, rejecting unknown fields and trailing
// data. Used for the fixed-shape ErrorPayload substructure.
func strictUnmarshal(raw json.RawMessage, v any) error {
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	dec.DisallowUnknownFields()
	if err := dec.Decode(v); err != nil {
		return err
	}
	if dec.More() {
		return errors.New("unexpected trailing data")
	}
	return nil
}

func requireString(fields map[string]json.RawMessage, name string) (string, error) {
	raw, ok := fields[name]
	if !ok {
		return "", newMalformed(fmt.Sprintf("missing field %q", name))
	}
	s, err := parseString(raw)
	if err != nil {
		return "", newMalformed(fmt.Sprintf("field %q must be a string", name))
	}
	if s == "" {
		return "", newMalformed(fmt.Sprintf("field %q must not be empty", name))
	}
	return s, nil
}

func requireInt64(fields map[string]json.RawMessage, name string) (int64, error) {
	raw, ok := fields[name]
	if !ok {
		return 0, newMalformed(fmt.Sprintf("missing field %q", name))
	}
	n, err := parseInt64(raw)
	if err != nil {
		return 0, newMalformed(fmt.Sprintf("field %q must be an integer", name))
	}
	return n, nil
}

func parseString(raw json.RawMessage) (string, error) {
	var s string
	if err := json.Unmarshal(raw, &s); err != nil {
		return "", err
	}
	return s, nil
}

// parseInt64 parses raw as an exact 64-bit integer, using json.Number so
// large values never pass through float64 (which loses precision above
// 2^53).
func parseInt64(raw json.RawMessage) (int64, error) {
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	var n json.Number
	if err := dec.Decode(&n); err != nil {
		return 0, err
	}
	return n.Int64()
}
